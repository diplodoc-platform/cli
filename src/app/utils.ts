import {TextSizes, Theme} from '@doc-tools/components';

const DEFAULT_USER_SETTINGS = {
    theme: Theme.Light,
    textSize: TextSizes.m,
    showMiniToc: true,
    wideFormat: true,
    fullScreen: false,
};

export function getDocSettings() {
    const {
        theme: defaultTheme,
        textSize: defaultTextSize,
        showMiniToc: defaultShowMiniToc,
        wideFormat: defaultWideFormat,
        fullScreen: defaultFullScreen,
    } = DEFAULT_USER_SETTINGS;

    const theme = getSetting('theme') as Theme || defaultTheme;
    const textSize = getSetting('textSize') as TextSizes || defaultTextSize;
    const showMiniToc = getSetting('showMiniToc') || defaultShowMiniToc;
    const wideFormat = getSetting('wideFormat') || defaultWideFormat;
    const fullScreen = getSetting('fullScreen') || defaultFullScreen;

    return {
        theme,
        textSize,
        showMiniToc: strToBoolean(showMiniToc),
        wideFormat: strToBoolean(wideFormat),
        fullScreen: strToBoolean(fullScreen),
    };
}

export function getSetting(name: string) {
    return sessionStorage.getItem(name);
}

export function saveSetting<T>(name: string, value: T) {
    sessionStorage.setItem(name, String(value));
}

export function strToBoolean(str: any) {
    if (typeof str === 'boolean') {
        return str;
    }

    return str ? str === 'true' : false;
}

export function withSavingSetting<T>(settingName: string, onChange: (value: T) => void) {
    return (value: T) => {
        saveSetting<T>(settingName, value);

        onChange(value);
    };
}

export function updateRootClassName(theme: Theme, isMobile = false) {
    const themeClassName = theme === 'light' ? 'yc-root_theme_light' : 'yc-root_theme_dark';
    const mobileClassName = isMobile ? 'mobile' : '';

    document.body.className = `yc-root ${themeClassName} ${mobileClassName}`;
}
