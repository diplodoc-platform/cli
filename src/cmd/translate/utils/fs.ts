import {JSONValue, resolveRefs} from '@diplodoc/translation';
import {dump, load} from 'js-yaml';
import {readFile, writeFile} from 'fs/promises';

function last<T>(array: T[]): T | undefined {
    return array[array.length - 1];
}

function ext(path: string) {
    const parts = path.split('.');

    if (last(parts) === 'skl') {
        parts.pop();
    }

    return last(parts);
}

function parseFile(text: string, path: string): JSONValue | string {
    if (typeof text !== 'string') {
        return text;
    }

    switch (ext(path)) {
        case 'yaml':
            return load(text) as object;
        case 'json':
            return JSON.parse(text);
        default:
            return text;
    }
}

function stringifyFile(content: JSONValue | string, path: string): string {
    if (typeof content === 'string') {
        return content;
    }

    switch (ext(path)) {
        case 'yaml':
            return dump(content);
        case 'json':
            return JSON.stringify(content);
        default:
            return content as unknown as string;
    }
}

export async function loadFile<T = string | JSONValue>(path: string, resolve = true): Promise<T> {
    const text = await readFile(path, 'utf8');

    let content = parseFile(text, path);

    if (content && typeof content === 'object' && resolve) {
        content = await resolveRefs(content, path, parseFile);
    }

    return content as T;
}

export async function dumpFile(path: string, content: string | JSONValue) {
    const text = stringifyFile(content, path);

    await writeFile(path, text, 'utf8');
}

/**
 * Takes toc schema if file matched as toc.
 * Takes leading schema if file matched as leading page.
 * Takes presets schema if file matched as presets.
 * Any way translation inner logic will search `$schema` attribute with high priority.
 * If `$schema` attribute not found anc precise schema not resolved,
 * we think that current yaml is a part of complex toc.yaml
 */
export async function resolveSchemas(path: string) {
    if (path.endsWith('toc.yaml')) {
        return [await loadFile('schemas/toc-schema.yaml', false)];
    }

    if (path.endsWith('index.yaml')) {
        return [await loadFile('schemas/leading-schema.yaml', false)];
    }

    if (path.endsWith('presets.yaml')) {
        return [await loadFile('schemas/presets-schema.yaml', false)];
    }

    return [await loadFile('schemas/toc-schema.yaml', false)];
}
