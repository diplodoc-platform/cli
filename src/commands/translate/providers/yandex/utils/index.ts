export {LimitExceed, RequestError, AuthError} from './errors';

export class Defer<T = string> {
    resolve!: (text: T) => void;

    reject!: (error: any) => void;

    promise: Promise<T>;

    constructor() {
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
    }
}

export function bytes(texts: string[]) {
    return texts.reduce((sum, text) => sum + text.length, 0);
}
