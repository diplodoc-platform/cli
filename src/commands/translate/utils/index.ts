export type {Locale} from './config';
export {resolveSchemas, FileLoader} from './fs';
export {extract, compose} from './translate';
export {resolveSource, resolveTargets, resolveFiles, resolveVars} from './config';
export {
    TranslateError,
    ExtractError,
    ComposeError,
    SkipTranslation,
    EmptyTokensError,
} from './errors';
