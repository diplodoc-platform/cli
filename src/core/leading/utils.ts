import {cloneDeepWith, isString} from 'lodash';

export function modifyValuesByKeys(
    object: object,
    keys: string[],
    modify: (value: string) => string,
) {
    // Clone the object deeply with a customizer function that modifies matching keys
    return cloneDeepWith(object, (value: unknown, key) => {
        if (keys.includes(key as string) && isString(value)) {
            return modify(value);
        }

        return undefined;
    });
}
