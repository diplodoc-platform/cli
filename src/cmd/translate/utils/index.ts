import {ok} from 'assert';
import {basename, dirname, extname, resolve} from 'path';
import {readFileSync} from 'node:fs';
import glob from 'glob';

export {dumpFile, loadFile, resolveSchemas} from './fs';
export {extract, compose} from './translate';
export {TranslateError, LimitExceed, RequestError, AuthError} from './errors';

type TranslateArgs = {
    input: string;
    output?: string;
    source?: string;
    sourceLanguage?: string;
    sourceLanguageLocale?: string;
    target?: string;
    targetLanguage?: string;
    targetLanguageLocale?: string;
    auth?: string;
    folder?: string;
    glossary?: string;
    include?: string[] | string;
    exclude?: string[] | string;
    dryRun?: boolean;
    useSource?: boolean;
} & {
    [prop: string]: any;
};

export type TranslateParams = {
    input: string;
    output: string;
    source: [string, string];
    targets: [string, string][];
    auth?: string;
    folder?: string;
    glossary?: string;
    files: string[];
    dryRun: boolean;
    useSource: boolean;
};

export function normalizeParams(params: TranslateArgs, exts = EXTS): TranslateParams {
    const source = normalizeLocale(params, 'source')[0];
    const targets = normalizeLocale(params, 'target');
    const {input, files} = normalizeInput(params, source[0], exts);
    const {output, auth, folder, dryRun, useSource} = params;

    ok(input, 'Required param input is not configured');

    return {
        input,
        output: output || input,
        auth,
        folder,
        source,
        targets,
        files: [...new Set(files)],
        dryRun: Boolean(dryRun),
        useSource: Boolean(useSource),
    };
}

function normalizeLocale(params: TranslateArgs, scope: 'source' | 'target'): [string, string][] {
    const _scopeLanguage = params[scope + 'Language'];
    const _scopeLanguageLocale = params[scope + 'LanguageLocale'];
    const _scope = params[scope] || _scopeLanguageLocale || _scopeLanguage;

    if (!_scope) {
        return [['', '']];
    }

    if (Array.isArray(_scope)) {
        return _scope.map((_scope) => _scope.split('-'));
    }

    return [_scope.split('-') as [string, string]];
}

const EXTS = ['.md', '.yaml', '.json'];

function normalizeInput(params: TranslateArgs, language: string, exts: string[]) {
    let {input, include = [], exclude = []} = params;

    include = ([] as string[]).concat(include);
    exclude = ([] as string[]).concat(exclude);

    let files: string[] | null = null;
    if (extname(input) === '.list') {
        const list = readFileSync(input, 'utf8').split('\n');
        input = dirname(input);
        files = list.map((file) => {
            const absPath = resolve(input, file);

            if (!absPath.startsWith(input)) {
                throw new Error(`Insecure access to file out of project scope. (file: ${absPath})`);
            }

            if (!exts.includes(extname(file))) {
                throw new Error(`Unhandles file extension. (file: ${absPath})`);
            }

            return absPath;
        });
    } else {
        if (exts.includes(extname(input))) {
            files = [basename(input)];
            input = dirname(input);
        }

        if (!include.length) {
            include.push('...');
        }

        include = include.reduce((acc, item) => {
            if (item === '...') {
                acc.push(...exts.map((ext) => (language || '.') + '/**/*' + ext));
            } else {
                acc.push(item);
            }

            return acc;
        }, [] as string[]);

        files =
            files ||
            ([] as string[]).concat(
                ...include.map((match) =>
                    glob.sync(match, {
                        cwd: input,
                        ignore: exclude,
                        nodir: true,
                    }),
                ),
            );
    }

    return {input, files};
}

export class Defer<T = string> {
    resolve!: (text: T) => void;

    reject!: (error: any) => void;

    promise: Promise<T>;

    constructor() {
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
    }
}

export function bytes(texts: string[]) {
    return texts.reduce((sum, text) => sum + text.length, 0);
}
