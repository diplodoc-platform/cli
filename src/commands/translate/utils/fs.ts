import type {JSONObject, LinkedJSONObject} from '@diplodoc/translation';
import {dirname, join, resolve} from 'node:path';
import {mkdir, readFile, writeFile} from 'node:fs/promises';
import {dump, load} from 'js-yaml';
import {linkRefs, unlinkRefs} from '@diplodoc/translation';

const ROOT = dirname(require.resolve('@diplodoc/cli/package'));

function last<T>(array: T[]): T | undefined {
    return array[array.length - 1];
}

function ext(path: string) {
    const parts = path.split('.');

    if (last(parts) === 'skl') {
        parts.pop();
    }

    return last(parts);
}

function parseFile(text: string, path: string): JSONObject | string {
    switch (ext(path)) {
        case 'yaml':
            return load(text) as JSONObject;
        case 'json':
            return JSON.parse(text);
        default:
            return text;
    }
}

function stringifyFile(content: JSONObject | string, path: string): string {
    if (typeof content === 'string') {
        return content;
    }

    switch (ext(path)) {
        case 'yaml':
            return dump(content, {noRefs: true});
        case 'json':
            return JSON.stringify(content);
        default:
            return content as unknown as string;
    }
}

function isObject(object: any): object is JSONObject {
    return object && typeof object === 'object';
}

export class FileLoader<T = string | JSONObject> {
    get data(): T {
        if (this._data === null) {
            throw new Error(`File content for ${this.path} is not loaded`);
        }

        return this._data;
    }

    get isString() {
        return typeof this.data === 'string';
    }

    get isObject() {
        return typeof this.data === 'object';
    }

    private _data: T | null = null;

    private parts: Record<string, T> = {};

    private path: string;

    private resolveRefs: boolean;

    constructor(path: string, resolveRefs = true) {
        this.path = path;
        this.resolveRefs = resolveRefs;
    }

    set(data: T) {
        this._data = data;
        this.parts[this.path] = this._data;

        return this;
    }

    async load() {
        const load = async (path: string, resolveRefs: boolean) => {
            path = resolve(path);

            if (!this.parts[path]) {
                const text = await readFile(path, 'utf8');
                const content = (this.parts[path] = parseFile(text, path) as T);

                if (isObject(content) && resolveRefs) {
                    await linkRefs(content, path, async (path: string) => {
                        if (!this.parts[path]) {
                            this.parts[path] = await load(path, false);
                        }

                        return this.parts[path] as JSONObject;
                    });
                }
            }

            return this.parts[path] as T;
        };

        if (!this._data) {
            this._data = await load(this.path, this.resolveRefs);
        }

        return this._data;
    }

    async dump(repath = (path: string) => path) {
        for (const path of Object.keys(this.parts)) {
            if (this.isObject && this.resolveRefs) {
                await unlinkRefs(this.parts[path] as LinkedJSONObject);
            }

            const output = repath(path);
            const text = this.isString
                ? (this.parts[path] as string)
                : stringifyFile(this.parts[path] as JSONObject, path);

            await mkdir(dirname(output), {recursive: true});
            await writeFile(output, text, 'utf8');
        }
    }
}

async function loadFile<T = string | JSONObject>(path: string): Promise<T> {
    return parseFile(await readFile(path, 'utf8'), path) as T;
}

/**
 * Takes toc schema if file matched as toc.
 * Takes leading schema if file matched as leading page.
 * Takes presets schema if file matched as presets.
 * Any way translation inner logic will search `$schema` attribute with high priority.
 * If `$schema` attribute not found anc precise schema not resolved,
 * we think that current yaml is a part of complex toc.yaml
 */
export async function resolveSchemas(path: string) {
    if (path.endsWith('toc.yaml')) {
        return [await loadFile(join(ROOT, 'schemas/toc-schema.yaml'))];
    }

    if (path.endsWith('index.yaml')) {
        return [await loadFile(join(ROOT, 'schemas/leading-schema.yaml'))];
    }

    if (path.endsWith('presets.yaml')) {
        return [await loadFile(join(ROOT, 'schemas/presets-schema.yaml'))];
    }

    if (path.endsWith('redirects.yaml')) {
        return [];
    }

    return [await loadFile(join(ROOT, 'schemas/toc-schema.yaml'))];
}
