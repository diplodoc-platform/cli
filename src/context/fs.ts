import {readFileSync, statSync, writeFileSync} from 'fs';
import {readFile, stat, writeFile} from 'fs/promises';
import {resolve} from 'path';
import {FsContext} from '@diplodoc/transform/lib/typings';
import {RevisionContext} from './context';

export function isFileExists(file: string) {
    try {
        const stats = statSync(file);

        return stats.isFile();
    } catch (e) {
        return false;
    }
}

export async function isFileExistsAsync(file: string) {
    try {
        const stats = await stat(file);

        return stats.isFile();
    } catch (e) {
        return false;
    }
}

export class FsContextCli implements FsContext {
    private context: RevisionContext;

    constructor(context: RevisionContext) {
        this.context = context;
    }

    getPaths(path: string) {
        const arr = [path];

        const isFromTmpInputFolder = path.startsWith(resolve(this.context.tmpInputFolder) + '/');
        if (isFromTmpInputFolder) {
            const assetPath = path.replace(resolve(this.context.tmpInputFolder) + '/', '');
            const originPath = resolve(this.context.userInputFolder, assetPath);

            arr.unshift(originPath);
        }

        return arr;
    }

    exist(path: string): boolean {
        const paths = this.getPaths(path);

        for (const path of paths) {
            if (isFileExists(path)) {
                return true;
            }
        }

        return false;
    }

    read(path: string): string {
        const paths = this.getPaths(path);

        for (const path of paths) {
            if (isFileExists(path)) {
                return readFileSync(path, 'utf8');
            }
        }

        throw Error(`File has not been found at: ${path}`);
    }

    write(path: string, content: string): void {
        writeFileSync(path, content, {
            encoding: 'utf8',
        });
    }

    async existAsync(path: string): Promise<boolean> {
        const paths = this.getPaths(path);

        for (const path of paths) {
            if (await isFileExistsAsync(path)) {
                return true;
            }
        }

        return false;
    }

    async readAsync(path: string): Promise<string> {
        const paths = this.getPaths(path);

        for (const path of paths) {
            if (await isFileExistsAsync(path)) {
                return await readFile(path, 'utf8');
            }
        }

        throw Error(`File has not been found at: ${path}`);
    }

    async writeAsync(path: string, content: string): Promise<void> {
        await writeFile(path, content, {
            encoding: 'utf8',
        });
    }
}
