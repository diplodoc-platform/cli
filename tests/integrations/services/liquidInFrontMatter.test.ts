import {readFile} from 'fs/promises';
import {parseExistingMetadata} from 'services/metadata/parse';
import {emplaceMetadata} from 'services/metadata/utils';

const propValuesMockPath = 'mocks/fileContent/metadata/substitutionsInMetadataPropertyValues.md';
const propKeysMockPath = 'mocks/fileContent/metadata/substitutionsInMetadataPropertyKeys.md';

describe('Front matter (metadata) transformations', () => {
    it('do not break when a property value contains Liquid-style variable substitutions', async () => {
        const fileContent = await readFile(propValuesMockPath, {encoding: 'utf-8'});

        const {metadata, metadataStrippedContent} = parseExistingMetadata(fileContent);
        const processedContent = emplaceMetadata(metadataStrippedContent, metadata);

        expect(metadata).toMatchSnapshot();
        expect(processedContent).toMatchSnapshot();
    });

    it('do not break when a property key contains Liquid-style variable substitutions', async () => {
        const fileContent = await readFile(propKeysMockPath, {encoding: 'utf-8'});

        const {metadata, metadataStrippedContent} = parseExistingMetadata(fileContent);
        const processedContent = emplaceMetadata(metadataStrippedContent, metadata);

        expect(metadata).toMatchSnapshot();
        expect(processedContent).toMatchSnapshot();
    });
});
