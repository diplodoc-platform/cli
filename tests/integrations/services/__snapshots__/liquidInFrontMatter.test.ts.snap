// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Front matter (metadata) transformations do not break when a property key contains Liquid-style variable substitutions 1`] = `
{
  "(({{ key-name }}))": "This one's key employs a different whitespace style in a substitution.",
  "(({{key-name}}))": "This one's key only consists of a substitution.",
  "(({{key-name}}))-prop": "This one's key starts with a substitution.",
  "prop-(({{key-name}}))": "This one has a substitution in a key name.",
}
`;

exports[`Front matter (metadata) transformations do not break when a property key contains Liquid-style variable substitutions 2`] = `
"---
prop-{{key-name}}: This one has a substitution in a key name.
{{key-name}}-prop: This one's key starts with a substitution.
{{key-name}}: This one's key only consists of a substitution.
{{ key-name }}: This one's key employs a different whitespace style in a substitution.
---
Blah.
"
`;

exports[`Front matter (metadata) transformations do not break when a property value contains Liquid-style variable substitutions 1`] = `
{
  "prop1": "This is a metadata property with a (({{substitution}})) in it.",
  "prop2": "This one contains (({{multiple}})) (({{substitutions}})).",
  "prop3": "This one has (({{substitutions}})) of (({{ different }})) (({{  styles}})).",
  "prop4": "This one has a (({{substitution}})) as well, but the string literal is single-quoted.",
  "prop5": "This one has no quotes at (({{all}})).",
  "prop6": "(({{this}})) starts with a substitution.",
  "prop7": "(({{this}})) one is a multiline (({{property}})).",
}
`;

exports[`Front matter (metadata) transformations do not break when a property value contains Liquid-style variable substitutions 2`] = `
"---
prop1: This is a metadata property with a {{substitution}} in it.
prop2: This one contains {{multiple}} {{substitutions}}.
prop3: This one has {{substitutions}} of {{ different }} {{  styles}}.
prop4: >-
  This one has a {{substitution}} as well, but the string literal is
  single-quoted.
prop5: This one has no quotes at {{all}}.
prop6: {{this}} starts with a substitution.
prop7: {{this}} one is a multiline {{property}}.
---
Blah.
"
`;

exports[`Front matter (metadata) transformations emit valid metadata when a variable is substituted with an ampty string 1`] = `
"---
prop1: 
prop2: 
prop3: 
prop4: 
---
"
`;
